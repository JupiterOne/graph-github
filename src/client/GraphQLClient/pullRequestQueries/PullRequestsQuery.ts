import { ResourceIteratee } from '../../../client';
import {
  PullRequestResponse,
  IteratePagination,
  CursorState,
  BaseQueryState,
  BuildQuery,
  RateLimitStepSummary,
} from '../types';
import utils from '../utils';
import { ExecutableQuery } from '../CreateQueryExecutor';
import fragments from '../fragments';

interface QueryState extends BaseQueryState {
  pullRequests: CursorState;
}

type QueryParams = {
  fullName: string;
  public: boolean;
  ingestStartDatetime: string;
  maxResourceIngestion: number;
  maxSearchLimit: number;
};

/**
 * Builds the leanest query possible
 * based on the provided queryState.
 * Pagination for sub-resources (commits, reviews, labels)
 * is performed separately.
 * @param queryParams
 * @param queryState
 */
export const buildQuery: BuildQuery<QueryParams, QueryState> = (
  queryParams,
  queryState,
): ExecutableQuery => {
  const query = `
      query (
        $issueQuery: String!, 
        $maxSearchLimit: Int!,
        $pullRequestsCursor: String
      ) {
        search(first: $maxSearchLimit, after: $pullRequestsCursor, type: ISSUE, query: $issueQuery) {
          issueCount
          edges {
            node {
              ${pullRequestFields(queryParams.public)}
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
        ...${fragments.rateLimit}
      }`;

  return {
    query,
    ...(queryState?.rateLimit && {
      rateLimit: queryState.rateLimit,
    }),
    queryVariables: {
      issueQuery: `is:pr repo:${queryParams.fullName} updated:>=${queryParams.ingestStartDatetime}`,
      maxSearchLimit: queryParams.maxSearchLimit,
      ...(queryState?.pullRequests?.hasNextPage && {
        pullRequestsCursor: queryState?.pullRequests.endCursor,
      }),
    },
  };
};

const pullRequestFields = (isPublicRepo: boolean) => {
  return `...on PullRequest {
    author {
      ...${fragments.teamMemberFields}
    }
    baseRefName
    baseRefOid
    baseRepository {
      name
      owner {
        ...on RepositoryOwner {
          login
        }
      }
    }
    body
    changedFiles
    createdAt
    databaseId
    headRefName
    headRefOid
    headRepository {
      name
      owner {
        ...on RepositoryOwner {
          login
        }
      }
    }
    id
    ${
      isPublicRepo
        ? `mergeCommit {
              ...on Commit {
                commitUrl
                oid
              }
              associatedPullRequests(first: 1) {
                nodes {
                  id
                  number
                }
              }
           }`
        : ''
    }
    merged
    mergedAt
    mergedBy {
      ...${fragments.teamMemberFields}
    }
    number
    reviewDecision
    state
    title
    updatedAt
    url
  }`;
};

/**
 * Processes the data, iterating over each pull request.
 * If inner resources of the pull request have multiple
 * pages, the pull request is added to a InnerResourceQuery queue
 * to be processed later.
 * @param responseData
 * @param iteratee
 * @param onInnerResourceQueryRequired
 * @param logger
 */
export const processResponseData = async (
  responseData: any,
  iteratee: ResourceIteratee<PullRequestResponse>,
): Promise<QueryState> => {
  if (!responseData) {
    throw new Error('responseData param is required.');
  }

  const rateLimit = responseData.rateLimit;
  const pullRequestEdges = responseData.search.edges;

  for (const edge of pullRequestEdges) {
    const pullRequest = edge.node;
    if (!utils.hasProperties(pullRequest)) {
      // If there's no data, pass - possible if permissions aren't correct in GHE
      continue;
    }

    await iteratee(utils.responseToResource(pullRequest));
  }

  return {
    rateLimit,
    pullRequests: responseData.search.pageInfo,
  };
};

/**
 * Using the provided params, Pull Requests are
 * iterated through.
 * Pagination is handled at the root and inner resource levels.
 * Inner resources include: commits, reviews, labels
 * See SinglePullRequestQuery.ts for more details.
 * @param queryParams
 * @param iteratee
 * @param execute
 * @param logger
 * @return Promise
 */
const iteratePullRequests: IteratePagination<
  QueryParams,
  PullRequestResponse
> = async (
  queryParams,
  execute,
  iteratee,
  logger,
): Promise<RateLimitStepSummary> => {
  const originalMaxSearchLimit = queryParams.maxSearchLimit;
  const requestLimitState = {
    isActive: false,
    count: originalMaxSearchLimit,
    limit: originalMaxSearchLimit,
  };
  let pullRequestFetched = 0;
  let queryCost = 0;
  let queryState: QueryState | undefined = undefined;
  let paginationComplete = false;

  const countIteratee = async (pullRequest: PullRequestResponse) => {
    pullRequestFetched++;
    if (requestLimitState.isActive) {
      requestLimitState.count--;
      if (requestLimitState.count === 0) {
        // reset state to continue requesting as default
        requestLimitState.isActive = false;
        requestLimitState.count = originalMaxSearchLimit;
        requestLimitState.limit = originalMaxSearchLimit;
        queryParams.maxSearchLimit = originalMaxSearchLimit;
        logger?.info(
          { queryParams, queryState },
          'Finish querying page by half the search limit.',
        );
      }
    }
    await iteratee(pullRequest);
  };

  while (!paginationComplete) {
    const executable = buildQuery(queryParams, queryState);

    let response: any;
    try {
      response = await execute(executable);
    } catch (err) {
      if (err.message?.includes('This may be the result of a timeout')) {
        requestLimitState.isActive = true;
        const newSearchLimit = Math.max(
          Math.floor(requestLimitState.limit / 2),
          1,
        );
        if (newSearchLimit === requestLimitState.limit) {
          // prevent infinite loop: newSearchLimit is 1 and it already failed using 1
          throw err;
        }
        requestLimitState.limit = newSearchLimit;
        queryParams.maxSearchLimit = newSearchLimit;
        logger?.info(
          { queryParams, queryState },
          'Search Pull Requests timeout. Start querying by half the search limit.',
        );
        continue;
      } else {
        throw err;
      }
    }

    queryState = await processResponseData(response, countIteratee);

    queryCost += queryState.rateLimit?.cost ?? 0;

    const exceededMaxResourceLimit =
      pullRequestFetched >= queryParams.maxResourceIngestion;

    paginationComplete =
      !queryState.pullRequests?.hasNextPage || exceededMaxResourceLimit;

    if (exceededMaxResourceLimit) {
      logger?.warn(
        {
          paginationComplete,
          pullRequestFetched,
          maxPullRequests: queryParams.maxResourceIngestion,
        },
        'Max PR resource ingestion was reached.',
      );
    }
  }

  return {
    totalCost: queryCost,
    limit: queryState?.rateLimit?.limit,
    remaining: queryState?.rateLimit?.remaining,
    resetAt: queryState?.rateLimit?.resetAt,
  };
};

export default { iteratePullRequests };
