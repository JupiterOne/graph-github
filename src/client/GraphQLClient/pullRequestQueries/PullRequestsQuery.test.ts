import PullRequestsQuery, {
  buildQuery,
  processResponseData,
} from './PullRequestsQuery';
import {
  emptyPullRequest,
  pullRequestsPublic,
  pullRequestsPublicInnerPagination,
  singleQueryFullResponse,
  singleQueryInnerResourcePaginationComplete,
} from './testResponses';

describe('PullRequestsQuery', () => {
  describe('#buildQuery', () => {
    test('first query - no query state', () => {
      // Act
      const executableQuery = buildQuery({
        fullName: 'J1-Test/musical-palm-tree',
        public: true,
        lastExecutionTime: '2011-10-05T14:48:00.000Z',
      });

      // Assert
      expect(executableQuery).toMatchSnapshot();
    });

    test('query with query state', () => {
      // Arrange
      const queryState = {
        pullRequests: { hasNextPage: true, endCursor: 'endCursor' },
      };

      // Act
      const executableQuery = buildQuery(
        {
          fullName: 'J1-Test/musical-palm-tree',
          public: true,
          lastExecutionTime: '2011-10-05T14:48:00.000Z',
        },
        queryState,
      );

      // Assert
      expect(executableQuery).toMatchSnapshot();
    });
  });

  describe('#processResponseData', () => {
    test('Pulling data out with no inner resource pagination', async () => {
      // Arrange
      const iteratee = jest.fn();
      const addToQueue = jest.fn();

      // Act
      const result = await processResponseData(
        pullRequestsPublic[0],
        iteratee,
        addToQueue,
      );

      // Assert
      expect(result).toMatchSnapshot();
      expect(iteratee).toHaveBeenCalledTimes(2);
      expect(iteratee.mock.calls[0][0]).toMatchSnapshot();
      expect(iteratee.mock.calls[1][0]).toMatchSnapshot();
      expect(addToQueue).not.toHaveBeenCalled();
    });
    test('Skips resources with inner resource pagination', async () => {
      // Arrange
      const iteratee = jest.fn();
      const addToQueue = jest.fn();

      // Act
      const result = await processResponseData(
        pullRequestsPublicInnerPagination,
        iteratee,
        addToQueue,
      );

      // Assert
      expect(result).toMatchSnapshot();
      expect(iteratee).toHaveBeenCalledTimes(0);
      expect(addToQueue).toHaveBeenCalledTimes(1);
    });
  });

  describe('#iteratePullRequests', () => {
    test('Pulling data out with no inner resource pagination', async () => {
      // Arrange
      const iteratee = jest.fn();
      const execute = jest
        .fn()
        .mockResolvedValueOnce(pullRequestsPublic[0])
        .mockResolvedValueOnce(pullRequestsPublic[1]);

      // Act
      const result = await PullRequestsQuery.iteratePullRequests(
        {
          fullName: 'J1-Test/happy-sunshine',
          public: true,
          lastExecutionTime: '2011-10-05T14:48:00.000Z',
        },
        execute,
        iteratee,
      );

      // Assert
      expect(result.totalCost).toBe(3);
      expect(iteratee).toHaveBeenCalledTimes(3);
      expect(iteratee.mock.calls[0][0]).toMatchSnapshot();
      expect(execute).toHaveBeenCalledTimes(2);
      expect(execute.mock.calls[0][0]).toMatchSnapshot();
      expect(execute.mock.calls[1][0]).toMatchSnapshot();
    });

    test('Pulling data out with inner resource pagination', async () => {
      // Arrange
      const iteratee = jest.fn();
      const execute = jest
        .fn()
        .mockResolvedValueOnce(pullRequestsPublicInnerPagination) // Entity is not used, but queued for single query
        .mockResolvedValueOnce(singleQueryFullResponse) // Part 1 of inner resources single query
        .mockResolvedValueOnce(singleQueryInnerResourcePaginationComplete) // Part 2 of inner resources single query
        .mockRejectedValue(
          new Error(
            'Pagination failed to stop! This response should never be reached.',
          ),
        );

      // Act
      const { totalCost } = await PullRequestsQuery.iteratePullRequests(
        {
          fullName: 'J1-Test/happy-sunshine',
          public: true,
          lastExecutionTime: '2011-10-05T14:48:00.000Z',
        },
        execute,
        iteratee,
      );

      // Assert
      expect(totalCost).toBe(3);
      expect(iteratee).toHaveBeenCalledTimes(1);
      expect(iteratee.mock.calls[0][0]).toMatchSnapshot();
      expect(execute).toHaveBeenCalledTimes(3);
      expect(execute.mock.calls[0][0]).toMatchSnapshot();
      expect(execute.mock.calls[1][0]).toMatchSnapshot();
      expect(execute.mock.calls[2][0]).toMatchSnapshot();
    });

    test('handle empty/partial responses', async () => {
      // Arrange
      const iteratee = jest.fn();
      const execute = jest.fn().mockResolvedValueOnce(emptyPullRequest[0]);

      // Act
      const { totalCost } = await PullRequestsQuery.iteratePullRequests(
        {
          fullName: 'J1-Test/happy-sunshine',
          public: true,
          lastExecutionTime: '2011-10-05T14:48:00.000Z',
        },
        execute,
        iteratee,
      );

      // Assert
      expect(totalCost).toBe(1);
      expect(iteratee).not.toHaveBeenCalled();
      expect(execute).toHaveBeenCalledTimes(1);
    });
  });
});
