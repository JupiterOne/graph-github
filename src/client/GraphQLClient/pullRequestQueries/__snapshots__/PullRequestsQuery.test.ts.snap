// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PullRequestsQuery #buildQuery first query - no query state 1`] = `
Object {
  "query": "
      query (
        $issueQuery: String!, 
        $maxSearchLimit: Int!,
        $maxLimit: Int!,
        $pullRequestsCursor: String
      ) {
        search(first: $maxSearchLimit, after: $pullRequestsCursor, type: ISSUE, query: $issueQuery) {
          issueCount
          edges {
            node {
              ...pullRequestFields
              
      ... on PullRequest {
        commits(first: $maxLimit) {
          totalCount
          nodes {
            commit {
              ...commitFields
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
      }
              
      ... on PullRequest {
        reviews(first: $maxLimit) {
          totalCount
          nodes {
            ...reviewFields
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
      }
              
    ... on PullRequest {
      labels(first: $maxLimit) {
        totalCount
        nodes {
          id
          name
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    } 
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
        ...rateLimit
      }",
  "queryVariables": Object {
    "issueQuery": "is:pr repo:J1-Test/musical-palm-tree updated:>=2011-10-05T14:48:00.000Z",
    "maxLimit": 100,
    "maxSearchLimit": 25,
  },
}
`;

exports[`PullRequestsQuery #buildQuery query with query state 1`] = `
Object {
  "query": "
      query (
        $issueQuery: String!, 
        $maxSearchLimit: Int!,
        $maxLimit: Int!,
        $pullRequestsCursor: String
      ) {
        search(first: $maxSearchLimit, after: $pullRequestsCursor, type: ISSUE, query: $issueQuery) {
          issueCount
          edges {
            node {
              ...pullRequestFields
              
      ... on PullRequest {
        commits(first: $maxLimit) {
          totalCount
          nodes {
            commit {
              ...commitFields
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
      }
              
      ... on PullRequest {
        reviews(first: $maxLimit) {
          totalCount
          nodes {
            ...reviewFields
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
      }
              
    ... on PullRequest {
      labels(first: $maxLimit) {
        totalCount
        nodes {
          id
          name
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    } 
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
        ...rateLimit
      }",
  "queryVariables": Object {
    "issueQuery": "is:pr repo:J1-Test/musical-palm-tree updated:>=2011-10-05T14:48:00.000Z",
    "maxLimit": 100,
    "maxSearchLimit": 25,
    "pullRequestsCursor": "endCursor",
  },
}
`;

exports[`PullRequestsQuery #iteratePullRequests Pulling data out with inner resource pagination 1`] = `
Object {
  "commits": Array [
    Object {
      "id": "commit1",
    },
  ],
  "id": "MDExOlB1bGxSZXF1ZXN0MTA=",
  "labels": Array [
    Object {
      "id": "label1=",
      "name": "bug",
    },
    Object {
      "id": "label2=",
      "name": "documentation",
    },
    Object {
      "id": "label3=",
      "name": "first-task",
    },
    Object {
      "id": "label4=",
      "name": "feature",
    },
  ],
  "reviews": Array [
    Object {
      "id": "review1",
    },
    Object {
      "id": "review2",
    },
    Object {
      "id": "review3==",
    },
  ],
  "title": "Update CHANGELOG.md",
}
`;

exports[`PullRequestsQuery #iteratePullRequests Pulling data out with inner resource pagination 2`] = `
Object {
  "query": "
      query (
        $issueQuery: String!, 
        $maxSearchLimit: Int!,
        $maxLimit: Int!,
        $pullRequestsCursor: String
      ) {
        search(first: $maxSearchLimit, after: $pullRequestsCursor, type: ISSUE, query: $issueQuery) {
          issueCount
          edges {
            node {
              ...pullRequestFields
              
      ... on PullRequest {
        commits(first: $maxLimit) {
          totalCount
          nodes {
            commit {
              ...commitFields
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
      }
              
      ... on PullRequest {
        reviews(first: $maxLimit) {
          totalCount
          nodes {
            ...reviewFields
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
      }
              
    ... on PullRequest {
      labels(first: $maxLimit) {
        totalCount
        nodes {
          id
          name
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    } 
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
        ...rateLimit
      }",
  "queryVariables": Object {
    "issueQuery": "is:pr repo:J1-Test/happy-sunshine updated:>=2011-10-05T14:48:00.000Z",
    "maxLimit": 100,
    "maxSearchLimit": 25,
  },
}
`;

exports[`PullRequestsQuery #iteratePullRequests Pulling data out with inner resource pagination 3`] = `
Object {
  "query": "
      query (
        $pullRequestNumber: Int!
        $repoName: String!
        $repoOwner: String!
        $maxCount: Int!
        $commitsCursor: String
        $reviewsCursor: String
        $labelsCursor: String
      ) {
          repository(name: $repoName, owner: $repoOwner) {
            pullRequest(number: $pullRequestNumber) {
              ...pullRequestFields
              
    commits(first: $maxCount, after: $commitsCursor) {
      totalCount
      nodes {
        commit {
          ...commitFields
        }
      }
      
      pageInfo {
        endCursor
        hasNextPage
      }
    }
              
    reviews(first: $maxCount, after: $reviewsCursor) {
      totalCount
      nodes {
        ...reviewFields
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }
              
    labels(first: $maxCount, after: $labelsCursor) {
      totalCount
      nodes {
        id
        name
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    } 
            }
          }
          ...rateLimit
      }",
  "queryVariables": Object {
    "maxCount": 100,
    "pullRequestNumber": undefined,
    "repoName": "musical-palm-tree",
    "repoOwner": "J1-Test",
  },
}
`;

exports[`PullRequestsQuery #iteratePullRequests Pulling data out with inner resource pagination 4`] = `
Object {
  "query": "
      query (
        $pullRequestNumber: Int!
        $repoName: String!
        $repoOwner: String!
        $maxCount: Int!
        
        $reviewsCursor: String
        $labelsCursor: String
      ) {
          repository(name: $repoName, owner: $repoOwner) {
            pullRequest(number: $pullRequestNumber) {
              ...pullRequestFields
              
              
    reviews(first: $maxCount, after: $reviewsCursor) {
      totalCount
      nodes {
        ...reviewFields
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }
              
    labels(first: $maxCount, after: $labelsCursor) {
      totalCount
      nodes {
        id
        name
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    } 
            }
          }
          ...rateLimit
      }",
  "queryVariables": Object {
    "labelsCursor": "Y3Vyc2==",
    "maxCount": 100,
    "pullRequestNumber": undefined,
    "repoName": "musical-palm-tree",
    "repoOwner": "J1-Test",
    "reviewsCursor": "Y3Vyc29yOnYyOpO0M",
  },
}
`;

exports[`PullRequestsQuery #iteratePullRequests Pulling data out with no inner resource pagination 1`] = `
Object {
  "closed": false,
  "commits": Array [
    Object {
      "id": "MDY6Q29tbWl0MjE6YmVkZDliNDIzNmY3NzNlMTAxNDIwYzVjNzU3OWIyMDNmODVkNmVkNw==",
    },
  ],
  "id": "MDExOlB1bGxSZXF1ZXN0MTA=",
  "labels": Array [
    Object {
      "id": "MDU6TGFiZWwxODE=",
      "name": "bug",
    },
    Object {
      "id": "MDU6TGFiZWwxODI=",
      "name": "documentation",
    },
    Object {
      "id": "MDU6TGFiZWwxODM=",
      "name": "duplicate",
    },
    Object {
      "id": "MDU6TGFiZWwxODU=",
      "name": "good first issue",
    },
  ],
  "reviews": Array [
    Object {
      "id": "MDE3OlB1bGxSZXF1ZXN0UmV2aWV3Mg==",
    },
    Object {
      "id": "MDE3OlB1bGxSZXF1ZXN0UmV2aWV3Mw==",
    },
  ],
  "title": "Update CHANGELOG.md",
}
`;

exports[`PullRequestsQuery #iteratePullRequests Pulling data out with no inner resource pagination 2`] = `
Object {
  "query": "
      query (
        $issueQuery: String!, 
        $maxSearchLimit: Int!,
        $maxLimit: Int!,
        $pullRequestsCursor: String
      ) {
        search(first: $maxSearchLimit, after: $pullRequestsCursor, type: ISSUE, query: $issueQuery) {
          issueCount
          edges {
            node {
              ...pullRequestFields
              
      ... on PullRequest {
        commits(first: $maxLimit) {
          totalCount
          nodes {
            commit {
              ...commitFields
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
      }
              
      ... on PullRequest {
        reviews(first: $maxLimit) {
          totalCount
          nodes {
            ...reviewFields
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
      }
              
    ... on PullRequest {
      labels(first: $maxLimit) {
        totalCount
        nodes {
          id
          name
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    } 
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
        ...rateLimit
      }",
  "queryVariables": Object {
    "issueQuery": "is:pr repo:J1-Test/happy-sunshine updated:>=2011-10-05T14:48:00.000Z",
    "maxLimit": 100,
    "maxSearchLimit": 25,
  },
}
`;

exports[`PullRequestsQuery #iteratePullRequests Pulling data out with no inner resource pagination 3`] = `
Object {
  "query": "
      query (
        $issueQuery: String!, 
        $maxSearchLimit: Int!,
        $maxLimit: Int!,
        $pullRequestsCursor: String
      ) {
        search(first: $maxSearchLimit, after: $pullRequestsCursor, type: ISSUE, query: $issueQuery) {
          issueCount
          edges {
            node {
              ...pullRequestFields
              
      ... on PullRequest {
        commits(first: $maxLimit) {
          totalCount
          nodes {
            commit {
              ...commitFields
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
      }
              
      ... on PullRequest {
        reviews(first: $maxLimit) {
          totalCount
          nodes {
            ...reviewFields
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
      }
              
    ... on PullRequest {
      labels(first: $maxLimit) {
        totalCount
        nodes {
          id
          name
        }
        pageInfo {
          endCursor
          hasNextPage
        }
      }
    } 
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
        ...rateLimit
      }",
  "queryVariables": Object {
    "issueQuery": "is:pr repo:J1-Test/happy-sunshine updated:>=2011-10-05T14:48:00.000Z",
    "maxLimit": 100,
    "maxSearchLimit": 25,
    "pullRequestsCursor": "Y3Vyc29yOjI=",
  },
}
`;

exports[`PullRequestsQuery #processResponseData Pulling data out with no inner resource pagination 1`] = `
Object {
  "pullRequests": Object {
    "endCursor": "Y3Vyc29yOjI=",
    "hasNextPage": true,
  },
  "rateLimit": Object {
    "cost": 2,
  },
}
`;

exports[`PullRequestsQuery #processResponseData Pulling data out with no inner resource pagination 2`] = `
Object {
  "closed": false,
  "commits": Array [
    Object {
      "id": "MDY6Q29tbWl0MjE6YmVkZDliNDIzNmY3NzNlMTAxNDIwYzVjNzU3OWIyMDNmODVkNmVkNw==",
    },
  ],
  "id": "MDExOlB1bGxSZXF1ZXN0MTA=",
  "labels": Array [
    Object {
      "id": "MDU6TGFiZWwxODE=",
      "name": "bug",
    },
    Object {
      "id": "MDU6TGFiZWwxODI=",
      "name": "documentation",
    },
    Object {
      "id": "MDU6TGFiZWwxODM=",
      "name": "duplicate",
    },
    Object {
      "id": "MDU6TGFiZWwxODU=",
      "name": "good first issue",
    },
  ],
  "reviews": Array [
    Object {
      "id": "MDE3OlB1bGxSZXF1ZXN0UmV2aWV3Mg==",
    },
    Object {
      "id": "MDE3OlB1bGxSZXF1ZXN0UmV2aWV3Mw==",
    },
  ],
  "title": "Update CHANGELOG.md",
}
`;

exports[`PullRequestsQuery #processResponseData Pulling data out with no inner resource pagination 3`] = `
Object {
  "closed": false,
  "commits": Array [
    Object {
      "id": "MDY6Q29tbWl0MjE6Y2FmNWNjYzA2MjYyMzAxYzFiNjg4NWJkYjc2NzllNTQ0NjVkZmQ5ZA==",
    },
  ],
  "id": "MDExOlB1bGxSZXF1ZXN0OQ==",
  "labels": Array [],
  "reviews": Array [],
  "title": "Create CHANGELOG.md #2",
}
`;

exports[`PullRequestsQuery #processResponseData Skips resources with inner resource pagination 1`] = `
Object {
  "pullRequests": Object {
    "endCursor": "Y3Vyc29yOjI=",
    "hasNextPage": false,
  },
  "rateLimit": Object {
    "cost": 1,
  },
}
`;
